lambdas/pkgpush/src/t4_lambda_pkgpush/__init__.py

```diff
@@ -285,50 +272,176 @@ def setup_user_boto_session(session):
 
 def auth(f):
     @functools.wraps(f)
-    def wrapper(request):
-        with setup_user_boto_session(get_user_boto_session(**get_user_credentials(request))):
-            return f(request)
+    def wrapper(event):
+        with setup_user_boto_session(get_user_boto_session(**get_user_credentials(event["credentials"]))):
+            return f(event["params"])
     return wrapper
 
 
-class ApiException(Exception):
-    def __init__(self, status_code, message):
-        super().__init__()
-        self.status_code = status_code
+class PkgpushException(Exception):
+    def __init__(self, message, data=None):
+        super().__init__(message, data)
         self.message = message
+        self.data = data
 
-    @classmethod
-    def from_botocore_error(cls, boto_error: ClientError):
-        boto_response = boto_error.response
-        status_code = boto_response['ResponseMetadata']['HTTPStatusCode']
-        message = "{0}: {1}".format(
-            boto_response['Error']['Code'],
-            boto_response['Error']['Message']
+    def asdict(self):
+        return {
+            "name": self.__class__.__name__,
+            "message": self.message,
+            "data": self.data,
+        }
+
+    def __str__(self):
+        return f"{self.__class__.__name__}: {self.message}"
+
+
+class S3HashLambdaUnhandledError(PkgpushException):
+    def __init__(self):
+        super().__init__("An error occured while calling S3 Hash Lambda.")
+
+
+class InvalidCredentials(PkgpushException):
```

Comment:
@sir-sigurd wdyt about having distinct exception classes for everything (vs having one class with assignable attributes e.g. name/code, message, data)?

Reply:
Having distinct exception classes seem to make more sense because it allows to catch the exact exception reason without further analysis of exception.